### PR TITLE
Feat: RWA Offboarding Checklist

### PR DESCRIPTION
**Pull request by amusingaxl**

### CODE REVIEW

File: spell/rwa-checklists.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    * [ ] `ilk` matches the format `RWAXYZ-A`
+    * [ ] Sanity Checks (constructor args, public vars, ... via `require` condition)
+    * [ ] `oracle.init`
+    * [ ] `vat.init.ilk`
+    * [ ] `jug.init.ilk` (this will set `duty` to `RAY` (zero pct) by default)
+    * [ ] `vat` rely `join`
+    * [ ] Increase Ilk Debt Ceiling (set DC + increase Global DC)
+    * [ ] File Ilk `pip` in the `spotter`
+    * [ ] File Ilk `mat` in the `spotter`
+    * [ ] Poke `spotter` to pull in the price
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply spell/rwa-checklists.md:100]: There's actually the correct way to calculate it. We should probably add it here.

File: spell/rwa-checklists.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    * [ ] `val` makes sense in context of the [rate mechanism](https://github.com/makerdao/developerguides/blob/master/mcd/intro-rate-mechanism/intro-rate-mechanism.md)
+  * [ ] IF multiple RWA ilks are being combined into one, `val` calculation is done once per ilk and added to make the total, with separate executable formulas provided in comments. The existing `val` value can be retrieved by calling `read()` on `PIP_RWAXYZ` and converting the result into decimal
+  * [ ] Oracle price is updated via `DssExecLib.updateCollateralPrice(ilk)`
+  * IF soft liquidation explicitly requested to be triggered (via `.tell(ilk)`) AND debt ceiling (`vat.ilks(ilk).line`) is `0` (OR is being set to `0` in the current spell)
+      * [ ] `RwaLiquidationOracle.tell(ilk)` call is present
+      * [ ] IF `RWAXYZ_A_INPUT_CONDUIT` is an instance of [`TinlakeMgr`](https://github.com/centrifuge/tinlake-maker-lib/blob/master/src/mgr.sol) (it is a Centrifuge integration), additional `TinlakeMgr.tell()` call is present (in order to prevent further `TIN` redemptions in the Centrifuge pool)
+
+## RWA Offboarding Checklist
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply spell/rwa-checklists.md:177]: The thing is that we usually don't remove anything from the ilk registry. I'm not sure if we should do it for this specific case only.

File: spell/rwa-checklists.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+* IF there is debt in the RWA Vault (`vat.urns(ilk, RWAXYZ_A_URN).art > 0`), proceed with the write-off (i.e.: vault is in default):
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply spell/rwa-checklists.md:179]: What do you mean by surplus? `ink > art`?

File: spell/rwa-checklists.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  * IF soft liquidation `.tell(ilk)` has **NOT** been called for the ilk in a previous spell:
+    * [ ] The value of `RwaLiquidationOracle.ilks(ilk).tau` MUST be zero, OTHERWISE the offboarding can only happen in a future spell.
+    * [ ] The debt ceiling (`vat.ilks(ilk).line`) is `0` OR is currently being set to `0`.
+    * [ ] `RwaLiquidationOracle.tell(ilk)` call is present
+    * [ ] IF `RWAXYZ_A_INPUT_CONDUIT` is an instance of [`TinlakeMgr`](https://github.com/centrifuge/tinlake-maker-lib/blob/master/src/mgr.sol) (it is a Centrifuge integration), additional `TinlakeMgr.tell()` call is present (in order to prevent further `TIN` redemptions in the Centrifuge pool)
+  * [ ] IF the stability fee for the ilk is not zero (`jug.ilks(ilk).duty > 1 * RAY`), `jug.drip(ilk)` call is present
+  * [ ] `RwaLiquidationOracle.cull(ilk, RWAXYZ_A_URN)` call is present
+* OTHERWISE IF there is no more debt remaining in the vault (`vat.urns(ilk, RWAXYZ_A_URN).art == 0`):
+  * [ ] The debt ceiling (`vat.ilks(ilk).line`) is currently being set to `0`.
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply spell/rwa-checklists.md:188]: RWAs gem joins are permissioned, so in theory there's no harm in leaving the permission there.
We should have a general checklist for "scuttling" contracts (aka the removal of all permissions in the system for a contract) that covers it.

File: spell/rwa-checklists.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (amusingaxl) [reply spell/rwa-checklists.md:100]: https://github.com/makerdao/pe-checklists/blob/1069a8ac98c6a450b95addabc31faf94cde3a2f8/archived/spell/spell-reviewer-goerli-checklist.md?plain=1#L189-L195

[the remark this answers is not in the document]
Comment (amusingaxl) [reply spell/rwa-checklists.md:102]: Please check [these](https://github.com/makerdao/pe-checklists/pull/42/files#diff-e736d7b55804471cd427e6b4bc3bf8a4f33850cb044cf9262a5fe02e6df1e7c9R88-R89).

File: spell/rwa-checklists.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+## RWA Offboarding Checklist
+
+* IF there is debt in the RWA Vault (`vat.urns(ilk, RWAXYZ_A_URN).art > 0`), proceed with the write-off (i.e.: vault is in default):
+  * IF soft liquidation `.tell(ilk)` has **NOT** been called for the ilk in a previous spell:
+    * [ ] The value of `RwaLiquidationOracle.ilks(ilk).tau` MUST be zero, OTHERWISE the offboarding can only happen in a future spell.
+    * [ ] The debt ceiling (`vat.ilks(ilk).line`) is `0` OR is currently being set to `0`.
+    * [ ] `RwaLiquidationOracle.tell(ilk)` call is present
+    * [ ] IF `RWAXYZ_A_INPUT_CONDUIT` is an instance of [`TinlakeMgr`](https://github.com/centrifuge/tinlake-maker-lib/blob/master/src/mgr.sol) (it is a Centrifuge integration), additional `TinlakeMgr.tell()` call is present (in order to prevent further `TIN` redemptions in the Centrifuge pool)
+  * [ ] IF the stability fee for the ilk is not zero (`jug.ilks(ilk).duty > 1 * RAY`), `jug.drip(ilk)` call is present
```
Comment (SidestreamStrongStrawberry) [inline spell/rwa-checklists.md:187]: Could there be a case `soft liquidation` has been called but `stability fee for the ilk is not zero`?

File: spell/rwa-checklists.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+  * [ ] Oracle price is updated via `DssExecLib.updateCollateralPrice(ilk)`
+  * IF soft liquidation explicitly requested to be triggered (via `.tell(ilk)`) AND debt ceiling (`vat.ilks(ilk).line`) is `0` (OR is being set to `0` in the current spell)
+      * [ ] `RwaLiquidationOracle.tell(ilk)` call is present
+      * [ ] IF `RWAXYZ_A_INPUT_CONDUIT` is an instance of [`TinlakeMgr`](https://github.com/centrifuge/tinlake-maker-lib/blob/master/src/mgr.sol) (it is a Centrifuge integration), additional `TinlakeMgr.tell()` call is present (in order to prevent further `TIN` redemptions in the Centrifuge pool)
+
+## RWA Offboarding Checklist
+
+* IF there is debt in the RWA Vault (`vat.urns(ilk, RWAXYZ_A_URN).art > 0`), proceed with the write-off (i.e.: vault is in default):
+  * IF soft liquidation `.tell(ilk)` has **NOT** been called for the ilk in a previous spell:
+    * [ ] The value of `RwaLiquidationOracle.ilks(ilk).tau` MUST be zero, OTHERWISE the offboarding can only happen in a future spell.
```
Comment (SidestreamStrongStrawberry) [inline spell/rwa-checklists.md:183]: Can the rest of checklist still happen except `RwaLiquidationOracle.cull(ilk, RWAXYZ_A_URN)`? (eg. `Set Debt ceiling to 0`) If then I'd prefer further specify to avoid confusion.

File: spell/rwa-checklists.md (hunk 5 above)
Comment (amusingaxl) [reply spell/rwa-checklists.md:187]: Yes... All Centrifuge vaults have non-zero stability fees, and some of them have been put under soft liquidation.

File: spell/rwa-checklists.md (hunk 6 above)
Comment (amusingaxl) [reply spell/rwa-checklists.md:183]: Yeah, it makes sense. I'll amend this.

File: spell/rwa-checklists.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+      * [ ] Executing formula locally provides integer number that matches the `val` in the spell
+    * [ ] `val` makes sense in context of the [rate mechanism](https://github.com/makerdao/developerguides/blob/master/mcd/intro-rate-mechanism/intro-rate-mechanism.md)
+  * [ ] IF multiple RWA ilks are being combined into one, `val` calculation is done once per ilk and added to make the total, with separate executable formulas provided in comments. The existing `val` value can be retrieved by calling `read()` on `PIP_RWAXYZ` and converting the result into decimal
+  * [ ] Oracle price is updated via `DssExecLib.updateCollateralPrice(ilk)`
+  * IF soft liquidation explicitly requested to be triggered (via `.tell(ilk)`) AND debt ceiling (`vat.ilks(ilk).line`) is `0` (OR is being set to `0` in the current spell)
+      * [ ] `RwaLiquidationOracle.tell(ilk)` call is present
+      * [ ] IF `RWAXYZ_A_INPUT_CONDUIT` is an instance of [`TinlakeMgr`](https://github.com/centrifuge/tinlake-maker-lib/blob/master/src/mgr.sol) (it is a Centrifuge integration), additional `TinlakeMgr.tell()` call is present (in order to prevent further `TIN` redemptions in the Centrifuge pool)
```
Comment (SidestreamStrongStrawberry) [inline spell/rwa-checklists.md:177]: This has been in the [`spell-reviewer-mainnet-checklist`](https://github.com/makerdao/pe-checklists/blob/5208017c2c5e04f9ec03817856e06ff9d0fe9ad1/spell/spell-reviewer-mainnet-checklist.md?plain=1#L212) but I'm unsure it was followed in the previous spells.
(eg. [soft liquidation for RWA005-A ](https://github.com/makerdao/spells-mainnet/blob/af7bcbfa8fbfd39981f47808caeb3fbed30dc865/archive/2023-09-27-DssSpell/DssSpell.sol#L104-L110))

Could you guide me through where I can find related code or help me understand why this is needed?

File: spell/rwa-checklists.md (hunk 6 above)
Comment (amusingaxl) [reply spell/rwa-checklists.md]: Please check now.

File: spell/rwa-checklists.md (hunk 7 above)
Comment (amusingaxl) [reply spell/rwa-checklists.md:177]: There's no official guide.
The Centrifuge Vaults are pretty much abandonware nowadays.
The TL;DR:
- Instead of the traditional input and output conduits, Centrifuge vaults have a component that belongs to them (even though they were written by people working for MakerDAO back then), dubbed [`TinlakeManager`](https://github.com/centrifuge/tinlake-maker-lib/blob/master/src/mgr.sol).
- The `TinlakeManager` instance associated with the vault fulfills both the role of the input and the output conduit.
- It also drives the interaction with the Centrifuge pools, so it can be seen as the adapter component that connects the Maker Protocol to the Centrifuge Protocol.
- It has some methods that are named the same as other RWA components, such as `wipe`, `lock`, `free`, `exit`, `draw` (from [`RwaUrn`](https://github.com/makerdao/rwa-toolkit/blob/master/src/urns/RwaUrn.sol)) and `tell`, `cull` (from [`RwaLiquidationOracle`](https://github.com/makerdao/rwa-toolkit/blob/master/src/oracles/RwaLiquidationOracle.sol)).
- `tell` and `cull` are **permissionless**, but such actions can only take place after the `RwaLiquidationOracle.tell()` and `RwaLiquidationOracle.cull()` have been called, respectively.
- It is important to call `tell` to prevent further redemptions in the related Centrifuge pool
  - It's not documented anywhere, but these calls were supposed to be handled by Centrifuge, however they haven't done it in the past and asked us to do it.
  - To prevent any unwanted behavior, we added this step to the checklist.
- As for `cull`, it's optional, as it only settles the state in `TinlakeManager`.
  - If an ilk was `cull`ed in `RwaLiquidationOracle`, technically there's nothing else that can be done through Centrifuge.

Comment (SidestreamStrongStrawberry) [reply spell/rwa-checklists.md:177]: Thank you for the explainer!

**Review by SidestreamStrongStrawberry — APPROVED** [review]

File: spell/rwa-checklists.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (oddaf) [reply spell/rwa-checklists.md:185]: I was meaning Ilk.Art < sum of all Urns Urn.art, but just realized that is impossible (they always move in unison). Resolving this one.

**Review by oddaf — APPROVED** [review]
